### ISSUE TITLE
Typescript gen : invalid SQL Function types

### ISSUE THREAD

**Issue by ngasull** [supabase/postgres-meta#842]
# Bug report

- [x] I confirm this is a bug with Supabase, not with my own application. 
- [x] I confirm I have searched the [Docs](https://docs.supabase.com), GitHub [Discussions](https://github.com/supabase/supabase/discussions), and [Discord](https://discord.supabase.com). 

## Describe the bug

For both input and input types, typescript generation of SQL functions is invalid in some cases :
- Primitive types are generated as non-nullable (`type` instead of `type | null`) even though the function may accept NULL as input or output
- Domain types are generated as `unknown`

I think that this bug is not caused by a PostgreSQL limitation as views produce the correct typing : primitives are nullable, domains generate as their underlying type and when domains are marked `NOT NULL` they are even non-nullable in typescript.

## To Reproduce

```sql
CREATE DOMAIN strict_text AS text NOT NULL;

CREATE FUNCTION some_function(arg strict_text)
RETURNS table (nulltext text, stricttext strict_text)
LANGUAGE SQL AS $$
   SELECT NULL::uuid, arg
$$;
```

Generated type with `supabase gen types typescript --local`
```typescript
export interface Database {
  public: {
    Functions: {
      some_function: {
        Args: {
          arg: unknown
        }
        Returns: {
          nulltext: string
          stricttext: unknown
        }[]
      }
    }
  }
}
```

## Expected behavior

Generated types should be
```typescript
export interface Database {
  public: {
    Functions: {
      some_function: {
        Args: {
          arg: string
        }
        Returns: {
          nulltext: string | null
          stricttext: string
        }[]
      }
    }
  }
}
```

## System information

- OS: [e.g. Linux NixOS]
- Version of supabase CLI: 1.50.8
- Version of Node.js: 18.15.0

## Additional context

I :heart: the work being done and the mindset at Supabase. I believe strict typing is crucial to the success of a "backend in the DB" and I could help with a PR if I'm being given some pointers to the code responsible of the TS generation :wave:

**Comment by Donnerstagnacht** [comment]
Is there an update on this @sweatybridge or @ngasull ? I think https://github.com/supabase/postgres-meta/issues/841 is a duplicate of this issue.

**Comment by ngasull** [comment]
@Donnerstagnacht PostgreSQL types are always nullable except domains which are NOT NULL.

Thus, I think this issue is broader than #841 : every returned type should be generated as `Type | null` (including enums) except for domains, which aren't implemented at all.

**Comment by Donnerstagnacht** [comment]
Right, makes sense. Thanks for the clarification.

May I ask, have you received feedback on your PR #573 or why is that one pending since 2 years?

And wouldn't it be a good first step to fix #841 and fix the issue for functions e.g. return base types and/or tables/composite types?

**Comment by ngasull** [comment]
Indeed, because it's such an impacting breaking change the issue falls into the technical design too. However because Supabase tries to push safety to the front of the scene, I really felt that handling nullability correctly was a priority.

Current choice of typing primitives as their NonNullable TS counterpart is a respectable design choice. It's convenient and avoids domains. However it's incorrect and there is no way around unsound types in current design.

I think that changing the null behavior for enums brings the same questions as for other types. Therefore, if I had to decide, I would tackle the issue as a whole in order to commit a single breaking change and to provide simpler and more consistent functionality (why would only enums be generated as nullable?)

No feedback came back to me since PR creation which, to be honest, is sadly the reason why I dropped my interest for Supabase. Unsafe function typing makes it not reliable enough to write business logic in my opinion.
